### ISSUE TITLE
Is there a way to listen for the second keystroke with timeout?

### ISSUE THREAD

**Issue by 0x78logs** [k0kubun/xremap#80]
Hello,

I would like to implement something that's similiar to how
key mappings are handled in vim.

say for example I want to remap `jk` to escape
instead of pressing `jk` all at once I can press `j` first then I have
a short amount of milliseconds to follow with `k`

is there a way to implement this using **Xremap**? and if yes can I control the timeout too?
*If the second key isn't presessed within the given timeout, that should cancel the listening*
thank you.

**Comment by k0kubun** [comment]
In your use case, do you want to remap both j→k and k→j as escape? 

If you meant j→k only, the closest thing that currently exists would be:

```yml
keymap:
  - remap:
      J:
        remap:
          K: Esc
```

which always translates a key sequence j→k to escape. We could possibly extend this to something like:

```yml
keymap:
  - remap:
      J:
        remap:
          K: Esc
        timeout_millis: 500
```

which results in:

* input: j → 500ms+ wait → k
  * output: j, k
* input: j → 500ms- wait → k
  * output: escape

Is this what you want?

**Comment by 0x78logs** [comment]
Hello,

I have tried the first snippet, `jk` will result in `Esc` however `j` will become useless I won't be able to type the letter j anywhere else.

the second snippet doesnt' work because `timeout_millis` isn't recognized!

the result must be something like the following

```
once j is pressed
  listen for 500ms if the user will follow with k ... if yes (k is pressed withing 500ms, convert to Esc)
  if 500ms passed by and user didn't register k, press j as normal (results in letter j)
```

Thank you.

**Comment by k0kubun** [comment]
> the second snippet doesnt' work because timeout_millis isn't recognized!

Did you read what I commented? I said "we could possibly extend this". That doesn't exist yet and I'm asking whether you want that or not. Could you please just answer Yes or No to what I asked?

> * input: j → 500ms+ wait → k
>   * output: 500ms+ wait → j → k
> * input: j → 500ms- wait → k
>   * output: 500ms- wait → escape
>
> Is this what you want?

**Comment by 0x78logs** [comment]
* input: j → 500ms+ wait → k
  * output: 500ms+ wait → j → k
  * No this isn't the correct behaviour

* input: j → 500ms- wait → k
   * output: 500ms- wait → escape
   * Yes that's the correct behaviour
   * Note: it should resolve to escape as soon as the user clicks on `k`
   * i.e. 500ms is the maximum time for listening.

if the user followed by a key that's not configured as a sequence
the timeout should get canceled, there's no need to wait
that way it would feel normal and not laggy

for example

* input: j → 500ms- wait → j
   * output: stop waiting → jj

* input: j → 500ms- wait → x
   * output: stop waiting → jx

I hope this explanation can help you. my best to you.

**Comment by k0kubun** [comment]
Ah, I got your point. I'm okay with making it behave that way when `timeout_millis` is specified.

Before moving forward, I'll need to think about how it should behave when `timeout_millis` is NOT specified. Let me investigate how Emacs or something similar behaves.

**Comment by 0x78logs** [comment]
we can make `timeout_millis` disabled by default, so that you can
enable it once you need it.

when it comes to implementation, I think if you make it work as a
listener who would listen for an X amount of time, cancel if there's
no response and act if there's a response within `timeout_millis`.

I think a reccursive solution to the problem would be best, because
if the key squence is more than two keystrokes things might get
complicated.

* `timeout_millis` must reset to its initial value if there's another
    keystroke that we should wait for.
* the sequence should get cancelled if the user don't respond,
    no matter how deep we're in.

however, it is such a powerful mapping system and I don't see it
implemented anywhere else other than vim and emacs ...
it enables us to make meaningful shortcuts, with verbs like
ciw ==> **c**hange **i**nside **w**ord

so fast to type, and it eliminates the need for ctrl, shift and all of
those akward keys.

**Comment by k0kubun** [comment]
> we can make timeout_millis disabled by default, so that you can
enable it once you need it.

My intention was that too in terms of the config. My concern was more about the behavior when it was intentionally disabled. We want to simplifythe implementation, and it's an option to change its behavior a bit if it makes more sense.

**Comment by k0kubun** [comment]
TBH it seems like we've discussed slightly different things all the time. Could you just try filing a pull request yourself? That would make the discussion a lot simpler.

**Comment by 0x78logs** [comment]
I wish if I could just make it easy and submit a PR, but I don't have
any experience with rust.

Instead I wrote something quickly with JS, so that the behavior is
clear.
You can test it yourself.

## User config
The user config below is as follow

* with `timeout_millis`
    `j => 10000ms wait => k => log Escape`
    `j => 10000ms wait => j => 10000ms wait => j => log Return`

* without `timeout_millis` (falling back to a default value)
    `m => 10000ms wait => m => log Mark`

* testing different `timeout_millis` in different nested levels
    `a => 3000ms => b => 2000ms => c => 1500ms => d => 250ms => log`

```
var config = {
    j: {
        k: function k() { log("Escape!!!"); },

        j: {
            j: function j() { log("Return!!!"); }
        },

        timeout_millis: 10000
    },

    m: {
        m: function m() { log("Mark!!!") }
    },

    a: {
        b: {
            c: {
                d: {
                    e: function e() { log("Nested sequence!!!"); },

                    timeout_millis: 250
                },

                timeout_millis: 1500
            },

            timeout_millis: 2000
        },

        timeout_millis: 3000
    }
};
```

## program logic

```
// aliases
var log = console.log;




// utils
function is_undefined(value) {
    return value === undefined;
}

function is_defined(value) {
    return !is_undefined(value);
}

function is_function(value) {
    return typeof value === "function";
}

function create_state(key, config, timeout = 500) {
    var sequence = config[key];
    timeout = sequence?.timeout_millis ?? timeout;

    return (
        is_undefined(sequence)
        ? undefined
        : { key, sequence, timeout }
    );
}

function reset() {
    return undefined;
}




// handle user key press events
var global = {
    state: undefined,
    timeout_id: 0
};

document.onkeypress = on_keypress;
function on_keypress(e) {
    global.state = (
        is_undefined(global.state)
        ? create_state(e.key, config)
        : create_state(
            e.key,
            global.state.sequence,
            global.state.timeout
        )
    );

    if (is_defined(global.state)) {
        log(`${global.state.timeout}ms before we reset your state.`);
        clearTimeout(global.timeout_id);

        global.timeout_id = setTimeout(
            function reset_state() {
                log("Too slow, resetting the state...");
                global.state = reset();
            },

            global.state.timeout
        );
    }

    if (is_defined(global.state) && is_function(global.state.sequence)) {
        clearTimeout(global.timeout_id);
        global.state.sequence();
        global.state = reset();
    }
}
```

`timeout_millis` are exagerrated just because of testing and making sure all is working,
in real word 100ms would be more than enough. so it won't feel laggy or anything

if you test it, you can let me know if you want me to explain something

**Comment by k0kubun** [comment]
I tried your code. It seems like every prefix key is just executed as soon as it's pressed, and you only demonstrated the behavior when it successfully reached the most nested configuration, `log("xxx!!!")`, which is sort of straightforward to me and I'm not concerned about that.

Could you revise your implementation to show what key events should be emitted? Just `console.log`-ing the key name is fine.

**Comment by 0x78logs** [comment]
###### "every prefix key is executed as soon as it's pressed"
Yes, that's how it is in vim and emacs

###### "only demonstrated the behavior when it is a success"
Yes, because that's the point if the pressed key is withing `.config` then yes there's a potential that the user
maybe is about to perform a key sequence. otherwise if the pressed key isn't specified in `.config` then silently ignore it
and don't even bother with it

###### which key events to be emmitted. 
`keypress` that's the only needed one here

**Comment by k0kubun** [comment]
> "every prefix key is executed as soon as it's pressed"
> Yes, that's how it is in vim and emacs

Are you serious? When I push m → m, what your implementation does is call `log("Mark!!!")` _in addition to_ emitting double `m` events where xremap currently does NOT emit any key event for that case and therefore is a breaking change. If you didn't intend it, please fix your implementation. I don't think you really intend it because you said things like _convert_ to Esc, etc. What you implemented is just "also" do xxx, not convert.

**Comment by 0x78logs** [comment]
Oh yes, `log("Mark!!!")` or any log is just a way to say here you can change to the desired key (i.e Esc et ... )

the implementation that I have wrote is nothing serious I just wrote it, so that we both are on the same page of the feature that could be added to xremap

According to what you said, Xremap rely on key events. for that case it is easy. Instead of `console` you can emit the whole key sequence thus Xremap will translate that sequence to an actual key change

I think it is easy and straightforward and there's no need for a breaking change

**Comment by k0kubun** [comment]
> I think it is easy and straightforward and there's no need for a breaking change

You only raised a partial requirement and I'm still missing the implication on the existing feature, so I still don't know whether it's easy or hard.

> the implementation that I have wrote is nothing serious I just wrote it, so that we both are on the same page of the feature that could be added to xremap

I'm sorry, but we're still not on the same page because you simply skipped the implementation of the part I'm interested in. Please do clarify every single thing that should actually happen, such as "m is pressed", using `console.log`. I cannot work on or review it before understanding what you're actually requesting. English has not helped us so far, so I'm expecting you to write them in code.

**Comment by 0x78logs** [comment]
Hello, I hope these console logs that I have added can help clarify the idea.

```Javascript
// user config
var config = {
    j: {
        k: function k() { log("!!!!! Congrats: Escape!!!"); },

        j: {
            j: function j() { log("!!!!! Congrats: Return!!!"); }
        },

        timeout_millis: 10000
    },

    m: {
        m: function m() { log("!!!!! Congrats: Mark!!!") }
    },

    a: {
        b: {
            c: {
                d: {
                    e: function e() { log("!!!!! Congrats: Nested sequence!!!"); },

                    timeout_millis: 250
                },

                timeout_millis: 1500
            },

            timeout_millis: 2000
        },

        timeout_millis: 3000
    }
};




// aliases
var log = console.log;




// utils
function compose(...functions) {
    return functions.reduce(compose_two);
}

function compose_two(f1, f2) {
    return get_arg;

    function get_arg(arg) {
        return f1(f2(arg));
    }
}

function is_undefined(value) {
    return value === undefined;
}

function is_defined(value) {
    return !is_undefined(value);
}

function is_function(value) {
    return typeof value === "function";
}

function create_state(key, config, timeout = 500) {
    var sequence = config[key];
    timeout = sequence?.timeout_millis ?? timeout;

    return (
        is_undefined(sequence)
        ? undefined
        : { key, sequence, timeout }
    );
}

function reset() {
    return undefined;
}

function is_object(value) {
    return value !== null && typeof value === "object";
}

function keys(obj) {
    return Object.keys(obj);
}

function object_create(key, value) {
    return { [key]: value };
}

function key_combos(obj) {
    return (
        keys(obj)
            .filter(not_eq("timeout_millis"))
            .map(object_to_string)
    );

    function object_to_string(key) {
        return (
            is_object(obj[key])
            ? concat(
                `${key} ==>`,
                key_combos(obj[key])
            )
            : key
        );
    }
}

function get_combos(target_confgi) {
    return key_combos(target_confgi);
}

function not_eq(value_one) {
    return get_value_two;

    function get_value_two(value_two) {
        return value_one !== value_two;
    }
}

function concat(a, collection) {
    return (
        collection
            .map(item => `${a} ${item}`)
            .join(", ")
    );
}

function prop(object) {
    return get_key;

    function get_key(key) {
        return object[key];
    }
}

function k(x) {
    return konstant;

    function konstant() {
        return x;
    }
}

function head(xs) {
    return xs[0];
}


// handle user key press events
var global = {
    state: undefined,
    timeout_id: 0,
    temp: {...config}
};

global.get_config = function(pressed_key) {
    return object_create(pressed_key, global.temp[pressed_key]);
}


document.onkeypress = on_keypress;
function on_keypress(e) {
    global.state = (
        is_undefined(global.state)
        ? create_state(e.key, config)
        : create_state(
            e.key,
            global.state.sequence,
            global.state.timeout
        )
    );

    if (is_defined(global.state)) {
        log(`You have pressed the key ${e.key}, our guess is you are
            about to perform one of the following key sequences:
            ${get_combos(global.get_config(e.key))}`);

        log(`NOTE: You have ${global.state.timeout}ms to perform
            the above key sequences, otherwise we will stop waiting
            for you and instead perform the normal behaviour of
            pressing ${e.key}. Which is just printing the letter
            "${e.key}" to the screen`);

        global.temp = global.temp[e.key];
        clearTimeout(global.timeout_id);
        global.timeout_id = setTimeout(
            function reset_state() {
                log(`Timeout, you either were too slow or you hit
                    "${e.key}" by mistake and didn't meant to perform
                    any combos that starts with "${e.key}".
                    when we listen for a potential combo that you
                    might perform, you won't be able to see
                    "${e.key}" printed to the screen unless you
                    timeout. So it might feel like the system isn't
                    responsive. you can get around that by reducing
                    "timeout_millis" in your config`);
                global.state = reset();
                global.temp = config;
            },

            global.state.timeout
        );
    }

    if (is_defined(global.state) && is_function(global.state.sequence)) {
        clearTimeout(global.timeout_id);
        global.state.sequence();
        global.state = reset();
    }

    log(`REMINDER: Pressing a key that isn't included in the config
        will be ignored, and as a result it will be typed immediately
        to the screen, which is its normal behaviour in the first
        place.`);
}

```

**Comment by k0kubun** [comment]
Is it intentional that I get a timeout log when I press `m`? Could you update your code to maintain backward compatibility?

**Comment by 0x78logs** [comment]
Yes it is intentional.
"m" `timeout_millis` is not specified inside the config for that reason it falls back to a default value of `500ms`
which is very quick, you have to double type "m" faster to finish the sequence otherwise the program will think that you just meant to type "m" with your keyboard

everytime you get a timeout log, in real life that should be translated to the normal behaviour of pressing that key but if the user succeeds to finish the sequence before the timeout then the program will perform the keybinding that are defined in the config (in our example we just have `log Congrats!!!!`)

**Comment by k0kubun** [comment]
I asked to maintain backward compatibility. IIRC Emacs's prefix key doesn't work that way at least by default, and one of xremap's major use cases is Emacs key emulation. So, when timeout isn't specified, I think it shouldn't have a timeout.

**Comment by 0x78logs** [comment]
Sounds great to me :)

**Comment by k0kubun** [comment]
I mean, I want you to describe the expected behavior by revising your code. That's the point of my question so far, and you haven't provided the whole picture of your proposal yet. My understanding is that this cannot be naturally implemented no matter what specification we'll choose, so I'd like to be sure about what you exactly expect.

**Comment by 0x78logs** [comment]
According to your previous comment, if timeout_millis isn't specefied then it won't have a timeout that means it will be cancelled immediately right ?
if I press `j` it will print `j` to the screen, there's no chance of performing a combo with no timeout

that sounds great, I don't see any problem!

**Comment by k0kubun** [comment]
Not cancelled. You should consider it like setting infinity to the timeout.

Because your description in English hasn't really made sense to me so far, I'd like to see your proposal in code. Even if you don't see problem, I still see the problem I've described in it.

So please post another version of code that handles missing timeout in a backward-compatible manner. I cannot implement anything until you propose something that makes sense.

**Comment by 0x78logs** [comment]
Oh, you want it to default to infinity if `timeout_millis` isn't defined!

I can change the above code easily to make it default to "Infinity" in the case of `timeout_millis` is missing, however I don't think that it would be useful at all.

if you think about it, we can't define "Infinity" so we would just choose a large timeout number which is like saying. Instead of defaulting to 500ms, default to 24h

do you think that would be useful ?

**Comment by k0kubun** [comment]
It shouldn't even be 24h or an arbitrarily large number, but it should just never timeout no matter how long you wait. It's just the Emacs behavior that xremap wants to support, so it's useful.

**Comment by 0x78logs** [comment]
I understand,

In this case thinking about the implimentation I will prefer to use a larger timeout number and let the program wait and listen for the user response.

better than setting an infinite timeout that will keep on restarting every second or so ... it will eat CPU and it is harder to manage

but at the end of the day they both have the same effect they will keep on waiting for the user to press a key

**Comment by k0kubun** [comment]
Forget an infinite timeout. I said it's "like" setting infinity. When there's no timeout, it just sleeps until the next key event and thus it doesn't consume CPU at all. Either way, please just provide a new JavaScript code. This issue is currently not actionable on my end, so I'll close this issue and reopen this once you provide code that describes what you're actually proposing. Thank you.

**Comment by 0x78logs** [comment]
why don't you just create a new branch and implement the feature the way you want, and I'll test it.

the feature is simple we want to be able to enter a sequence of keys to perform an action, do it any way you want while supporting the emacs behaviour and it will be just fine

**Comment by k0kubun** [comment]
If the feature is really simple, why don't you just implement it? And I'll test and merge it.

**Comment by k0kubun** [comment]
I'll give up asking you to write code. I'll just design the feature and implement one myself. However, please note that it's going to take time for me to find time since I get tens of GitHub notifications like this a week and this requires a relatively larger effort than other ones to come up with a design that works well with the existing implementation. I'll likely have more free time in May, so if you can wait until May, you don't need to write code. I'll leave this issue open.

**Comment by 0x78logs** [comment]
I honestly just didn't know what exactly you didn't understand from my proposal but that's okay.

what do you mean by merging? if you can call a nodeJS module then I'll write something serious at the end of this month and I'll try to make it 100% independant

**Comment by k0kubun** [comment]
Sorry, I probably confused you by using the wrong word. I just wanted to see a working logic that keeps backward compatibility (your code didn't) and achieves what you want, and then I would rewrite that to Rust. 

Anyway, please just wait until May.

**Comment by zach-is-my-name** [comment]
I've been trying to publicize this repo because I think it's accomplishing incredible things that have apparently been too difficult to accomplish for the Linux community.  Namely making a powerful yet simple to use keymapper.  Options exist fulfilling one, but not both.  I am anxiously awaiting this feature and predict, STRONG adoption of xremap if it gets enough exposure

**Comment by k0kubun** [comment]
Thank you for the kind comment. To show my appreciation for you being my GitHub Sponsor, I filed https://github.com/k0kubun/xremap/pull/82.

This issue has not explained what's the expected behavior, so this is an implementation of my random guess. Please leave a comment to the pull request with a config you'd like to use and explain the expected behavior using the format of https://github.com/k0kubun/xremap/pull/82#issue-1189348666 if it doesn't work as expected.

**Comment by zach-is-my-name** [comment]
> Thank you for the kind comment. To show my appreciation for you being my GitHub Sponsor, I filed #82.
> 
> This issue has not fully explained what's the expected behavior, so this is an implementation of my random guess. Please leave a comment to the pull request with a config you'd like to use and explain the expected behavior using the format of [#82 (comment)](https://github.com/k0kubun/xremap/pull/82#issue-1189348666) if it doesn't work as expected.

I see that.. thank you ask well.  Most of your new stars probably came from Reddit  ; )

**Comment by k0kubun** [comment]
Released as v0.3.2.
